lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -206,44 +234,62 @@ defmodule Mix.Compilers.Elixir do
   # files that have changed. It then, recursively, figures out
   # all the files that changed (via the module dependencies) and
   # return the non-changed entries and the removed sources.
-  defp update_stale_entries(all, [], stale) when stale == %{} do
-    {all, []}
+  defp update_stale_entries(modules, files, [], stale) when stale == %{} do
+    {modules, files, []}
   end
 
-  defp update_stale_entries(all, changed, stale) do
+  defp update_stale_entries(modules, files, changed, stale) do
     removed = Enum.into(changed, %{}, &{&1, true})
-    remove_stale_entries(all, stale, removed)
+    remove_stale_entries(modules, files, stale, removed)
   end
 
-  defp remove_stale_entries(entries, old_stale, old_removed) do
-    {rest, new_stale, new_removed} =
-      Enum.reduce entries, {[], old_stale, old_removed}, &remove_stale_entry/2
+  defp remove_stale_entries(modules, files, old_stale, old_removed) do
+    file_modules =
+      Enum.reduce modules, %{}, fn module(source: source) = module, file_modules ->
```

Comment:
Can we put parens around the arguments for the `fn`? It looks pretty wild without them IMO, the things on this line all look like potential arguments to `Enum.reduce`. Wdyt? :)